analytics/analytics/job_processor/prometheus.py

```diff
@@ -0,0 +1,293 @@
+
+        params = {
+            "query": query,
+            "start": start.timestamp(),
+            "end": end.timestamp(),
+            "step": step,
+        }
+        query_params = urlencode(params)
+        query_url = f"http://{self.api_url}/query_range?{query_params}"
+        res = requests.get(query_url)
+        res.raise_for_status()
+
+        return res.json()["data"]["result"]
+
+    def annotate_resource_requests_and_limits(self, job: Job):
+        """Annotate cpu and memory resource requests and limits."""
+        pod = job.pod.name
+
+        def extract_value(result: dict | None) -> int | float | None:
+            if result is None:
+                return None
+
+            num = float(result["value"][1])
+            if num.is_integer():
+                num = int(num)
+
+            return num
+
+        # list where one entry is cpu, the other is mem
+        resource_requests = self.query_single(
+            f"kube_pod_container_resource_requests{{container='build', pod='{pod}'}}",
+            time=job.midpoint,
+        )
+        job.pod.cpu_request = extract_value(
+            next(
+                (rr for rr in resource_requests if rr["metric"]["resource"] == "cpu"),
+                None,
+            )
+        )
+        job.pod.memory_request = extract_value(
+            next(
+                (
+                    rr
+                    for rr in resource_requests
+                    if rr["metric"]["resource"] == "memory"
+                ),
+                None,
+            )
+        )
+
+        # list where one entry is cpu, the other is mem
+        resource_limits = self.query_single(
+            f"kube_pod_container_resource_limits{{container='build', pod='{pod}'}}",
+            time=job.midpoint,
+        )
+        job.pod.cpu_limit = extract_value(
+            next(
+                (rr for rr in resource_limits if rr["metric"]["resource"] == "cpu"),
+                None,
+            )
+        )
+        job.pod.memory_limit = extract_value(
+            next(
+                (rr for rr in resource_limits if rr["metric"]["resource"] == "memory"),
+                None,
+            )
+        )
+
+    def annotate_usage_and_occupancy(self, job: Job):
+        node = job.node.name
+        pod = job.pod.name
+
+        # Step is seconds between samples
+        step = 30
+
+        # Get cpu seconds usage
+        results = self.query_range(
+            f"container_cpu_usage_seconds_total{{container='build', node='{node}'}}",
+            start=job.started_at,
+            end=job.finished_at,
+            step=step,
+        )
+
+        # First, get the cpu utlization by the pod we care about
+        # To do this, just get the last value from the response, since that'll be the total of the counter
+        pod_results = next(
+            (res for res in results if res["metric"]["pod"] == pod), None
+        )
+        assert pod is not None, f"Pod {pod} not found in query"
```

Comment:
I was recently scolded for using `assert` instead of an if statement, rightfully so I think 😞.

Reply:
[2fb0190](https://github.com/spack/spack-infrastructure/pull/723/commits/2fb0190ef37d54034200acfd82211f98ba8733e8)